Development feat#65

tomski80 opened the pull request:
User can only add one empty card. If try to add more, then will see the message to edit empty card before adding more cards. If last card is left empty and user navigate away from edit component this card will be removed. (so there will not be empty cards in collection).

tomski80 committed:
feat: prevent user from adding empty cards

tomski80 committed:
fix: remove unnecessary console.log

AntonTymczenko commented on src/components/CollectionEdit.vue:
```diff
@@ -19,6 +19,7 @@
     class="newCard"
     @click="add"
   >+</button>
+  <p v-if=emptyWarningMessage>Please edit <em>Question</em> and <em>Answer</em> before adding new Card</p>
```
v-if="..."
you've missed double quotes around the variable

AntonTymczenko committed:
feat: Add on blur method

AntonTymczenko committed:
feat: Add two computed properties

AntonTymczenko committed:
feat: Use computed `lastIndex`

AntonTymczenko committed:
feat: Use two new computed props and auto check last empty as an error

This also (like before) keeps user from making two new empty cards in
the end, but in a simplier way

AntonTymczenko committed:
feat: Blur method

AntonTymczenko committed:
feat: Prevent route change if there are some errors

Still allows to change route if only the last card is empty

AntonTymczenko committed:
feat: More elaborate error messages.

Should be improved so that specific input fields are highlighted with
red color.

AntonTymczenko committed:
style: Add comment

AntonTymczenko committed:
feat: Simplify error message text rendering

AntonTymczenko committed:
feat: Show error about last card only if it is partitially filled

AntonTymczenko committed:
feat: `lastCardIsEmpty` prop is computed

AntonTymczenko committed:
fix: Reverse the `blur` logic to fix bugs with last card

AntonTymczenko commented:
oh, there's still one bug. 
when user makes an empty field in some card and then deletes this card -- an error is still in "errors" array. 

so we should remove errors with this index when deleting a card

tomski80 commented:
I think you over-engineered this now. It doesn't show the message now if I try to add another empty card that you have to fill in q & a before adding new empty - and in my opinion it is confusing, because button suddenly doesn't work . Also message to fill question and answer show when I start to writing either question or answer - so it is no longer empty (have at least one character in it) - so user obviously already doing it. Which is confusing as well. I think we need to rethink What we trying to communicate to the user here.

svmi3195 commented:
You make a code for testing if the card is empty
`    lastCardIsEmpty () {`
`      const lastCard = this.collection.items[this.lastIndex]`
`      return lastCard.q === '' && lastCard.a === ''`
`    }`
but you don't use it. I think you could use it here:
`    removeLastEmpty(){`
`      if(this.collection.items[this.lastIndex].q === ''`
`        && this.collection.items[this.lastIndex].a === '')`
`      {`
`        this.collection.items.pop()`
`      }`
`    }`

AntonTymczenko committed:
Merge branch 'development' into development-feat#65

AntonTymczenko commented:
@svmi3195 nicely noticed! I've put this computed property here too!

AntonTymczenko commented:
@tomski80 you are right, it is still buggy, but we can improve it :D I'll think about the solution in a few nearest hours...

AntonTymczenko committed:
feat: Update error list when removing a card

AntonTymczenko committed:
fix: Use computed prop

AntonTymczenko committed:
fix: Empty last card is not an error

AntonTymczenko committed:
feat: Show all errors

AntonTymczenko committed:
feat: Check last empty before add another one empty

AntonTymczenko committed:
fix: Don't add new empty card if in previous only one field was filled

AntonTymczenko committed:
refactor: `lastCardIsNotFilled === 'both'` instead of `lastCardIsEmpty`

AntonTymczenko committed:
fix: Remove 'ok to save' message

AntonTymczenko committed:
feat: Red inputs instead of error messages
